File: 11_Heap/c/heap.c
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

#define MAX_SIZE 100

typedef struct {
    int data[MAX_SIZE];
    int size;
} MaxHeap;

void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

void insert(MaxHeap *h, int val) {
    if (h->size >= MAX_SIZE) return;
    
    h->data[h->size] = val;
    int current = h->size;
    h->size++;
    
    while (current > 0) {
        int parent = (current - 1) / 2;
        if (h->data[current] > h->data[parent]) {
            swap(&h->data[current], &h->data[parent]);
            current = parent;
        } else {
            break;
        }
    }
}

int getMax(MaxHeap *h) {
    return (h->size > 0) ? h->data[0] : -1;
}

int main() {
    MaxHeap h;
    h.size = 0;
    
    insert(&h, 10);
    insert(&h, 30);
    insert(&h, 20);
    insert(&h, 50);
    
    assert(getMax(&h) == 50);
    printf("Test C Max-Heap réussi !\n");
    return 0;
}

File: 11_Heap/python/heap.py
class MaxHeap:
    def __init__(self):
        self.heap = []

    def insert(self, val):
        self.heap.append(val)
        self._percolate_up(len(self.heap) - 1)

    def get_max(self):
        return self.heap[0] if self.heap else None

    def _percolate_up(self, index):
        parent = (index - 1) // 2
        if index > 0 and self.heap[index] > self.heap[parent]:
            self.heap[index], self.heap[parent] = self.heap[parent], self.heap[index]
            self._percolate_up(parent)

if __name__ == "__main__":
    max_heap = MaxHeap()
    max_heap.insert(10)
    max_heap.insert(30)
    max_heap.insert(20)
    max_heap.insert(50)  # Le plus grand
    
    assert max_heap.get_max() == 50, "Erreur: Le max devrait être 50"
    print("Test Python Max-Heap réussi !")

